source/byte_buf.c

```diff
@@ -651,9 +651,9 @@ static int s_aws_byte_buf_append_dynamic(
     bool clear_released_memory) {
     AWS_PRECONDITION(aws_byte_buf_is_valid(to));
     AWS_PRECONDITION(aws_byte_cursor_is_valid(from));
-    AWS_ERROR_PRECONDITION(to->allocator);
 
     if (to->capacity - to->len < from->len) {
+        AWS_ERROR_PRECONDITION(to->allocator);
```

Comment:
if you're doing this to protect against `to == NULL`
that what the earlier asserts are for.
passing NULL into any of our functions (except cleanup functions) is undefined behavior, we don't try to defend against except by adding asserts

Reply:
reverted this in next rev. 
its not about to being potentially NULL, but more about trying to make behavior of zeroed out buffer more consistent across append apis. In retrospect through i agree it was not the best idea.
slight rant: corner cases for zeroed buffers do get a bit confusing, ex append_and_update will fail if buffer is zeroed out, unless the cursor has length zero as well, in which case our cursor is replaced with zeroed out cursor.

Reply:
we should treat a zeroed-out buffer as a non-dynamic buffer of 0 length.
so any "dynamic" functions should immediately reject it because it doesn't have an allocator and can't resize

non-dynamic append functions should handle them though ... I guess trying to add a 0-length cursor to a zeroed-out-buffer should succeed, but trying to append >0 bytes should fail

Reply:
agreed these APIs still have some footguns
the current rules are:
- it's an error to pass a "static" buffer to any "dynamic" function
- the non-dynamic functions just check if the new data would "fit". So even if your buffer is "full", it's OK if you try to append something of length 0.

If I could go back in time ...
- 100% different APIs for static and dynamic buffers, so it's impossible to pass a static buffer to a dynamic append funtion
- the dynamic API would be full of void functions, because allocation failure is fatal